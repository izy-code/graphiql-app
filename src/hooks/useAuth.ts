import { useContext } from 'react';

import type { AuthContextType } from '@/contexts/auth-context';
import { AuthContext } from '@/contexts/auth-context';

export const useAuth = (): AuthContextType => {
  const context = useContext(AuthContext);

  if (context === undefined) {
    throw new Error('useAuth can only be used inside AuthProvider');
  }

  return context;
};
