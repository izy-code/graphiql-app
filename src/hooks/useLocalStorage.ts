import { useCallback } from 'react';

import type { LocalStorageKeys } from '@/common/enums';

export const LOCAL_STORAGE_KEY = 'graphiql-adventurers-team';

export function getLocalStorage<T>(): Record<string, T> {
  const item = localStorage.getItem(LOCAL_STORAGE_KEY);
  return item ? (JSON.parse(item) as Record<string, T>) : {};
}

export function useLocalStorage<T>(): {
  getStoredValue: (key: LocalStorageKeys) => T | null;
  setStoredValue: (key: LocalStorageKeys, value: T) => void;
} {
  const getStoredValue = useCallback((key: LocalStorageKeys): T | null => getLocalStorage<T>()[key] ?? null, []);

  const setStoredValue = useCallback((key: LocalStorageKeys, value: T) => {
    let localStorageMap = getLocalStorage<T>();

    localStorageMap = { ...localStorageMap, [key]: value };

    localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(localStorageMap));
  }, []);

  return { getStoredValue, setStoredValue };
}
