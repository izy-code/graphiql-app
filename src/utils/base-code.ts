export const encodeBase64 = (str: string): string => Buffer.from(str, 'utf-8').toString('base64');
export const decodeBase64 = (str: string): string => Buffer.from(str, 'base64').toString('utf-8');
