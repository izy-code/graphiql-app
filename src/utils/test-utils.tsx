import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { type PropsWithChildren, type ReactElement, useRef } from 'react';
import { Provider } from 'react-redux';

import type { AppStore, RootState } from '@/store/store';
import { setupStore } from '@/store/store';

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

export function renderWithProvidersAndUser(
  ui: ReactElement,
  { preloadedState = {}, store = setupStore(preloadedState), ...renderOptions }: ExtendedRenderOptions = {},
): { store: AppStore; user: ReturnType<typeof userEvent.setup>; container: HTMLElement } {
  function Wrapper({ children }: PropsWithChildren): ReactElement {
    const storeRef = useRef<AppStore>();

    if (!storeRef.current) {
      storeRef.current = setupStore(preloadedState);
    }

    return <Provider store={storeRef.current}>{children}</Provider>;
  }

  return { store, user: userEvent.setup(), ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithUserSetup(ui: ReactElement): {
  user: ReturnType<typeof userEvent.setup>;
  container: HTMLElement;
} {
  return {
    user: userEvent.setup(),
    ...render(ui),
  };
}
