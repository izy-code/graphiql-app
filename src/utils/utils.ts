import {
  LOWERCASE_LETTER_REGEX,
  MIN_PASSWORD_LENGTH,
  NUMBER_REGEX,
  SPECIAL_CHARACTER_REGEX,
  UPPERCASE_LETTER_REGEX,
} from '@/common/constants';
import type { TableRow } from '@/components/client-table/types.ts';
import en from '@/locales/en';
import ru from '@/locales/ru';

const regexArray = [NUMBER_REGEX, UPPERCASE_LETTER_REGEX, LOWERCASE_LETTER_REGEX, SPECIAL_CHARACTER_REGEX];
export const MAX_STRENGTH = regexArray.length + 1;
export const getPasswordStrength = (password: string): number => {
  let strength = regexArray.reduce((result, regex) => result + Number(regex.test(password)), 0);
  strength += password.length >= MIN_PASSWORD_LENGTH ? 1 : 0;

  return strength;
};

export const translateText = (key: string): string =>
  window.location.pathname.split('/').includes('en') ? en[key as never] : ru[key as never];

export const encodeBase64 = (str: string): string => Buffer.from(str, 'utf-8').toString('base64').replace(/=+$/, '');

export const decodeBase64 = (str: string): string => {
  const pad = str.length % 4 === 0 ? '' : '===='.slice(str.length % 4);
  return Buffer.from(str + pad, 'base64').toString('utf-8');
};

export const generateUniqueId = (): string => `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;

export const replaceVariables = (text: string | undefined, variables: TableRow[]): string => {
  const variableMap = Object.fromEntries(
    variables.filter(({ key, value }) => key.trim() && value.trim()).map(({ key, value }) => [key, value]),
  );

  let result = text || '';
  Object.entries(variableMap).forEach(([variableKey, variableValue]) => {
    const variablePlaceholder = `{{${variableKey}}}`;
    result = result.replace(new RegExp(variablePlaceholder, 'g'), variableValue);
  });

  return result.trim();
};

export const getEncodedHeaders = (headersParameter: TableRow[]): string => {
  if (!headersParameter || headersParameter.length === 0) {
    return '';
  }

  const encodedHeaders = headersParameter
    .filter(({ key, value }) => key.trim() && value.trim())
    .map(({ key, value }) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`);

  return `?${encodedHeaders.join('&')}`;
};

export const updateHistory = (
  locale: string,
  method: string,
  endpoint: string,
  body: string,
  headers: TableRow[],
  variables: TableRow[],
): void => {
  const encodedEndpoint = endpoint ? encodeBase64(endpoint) : encodeBase64(' ');
  const encodedBody = encodeBase64(replaceVariables(body, variables) || ' ');
  const encodedHeaders = getEncodedHeaders(headers);

  window.history.replaceState(null, '', `/${locale}/${method}/${encodedEndpoint}/${encodedBody}${encodedHeaders}`);
};
