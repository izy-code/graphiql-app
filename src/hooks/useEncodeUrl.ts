import { useCallback } from 'react';

import { NO_ENDPOINT } from '@/common/constants';
import { ProtectedPaths } from '@/common/enums';
import type { IData } from '@/components/client-table/types';
import { useCurrentLocale } from '@/locales/client';
import type { RootState } from '@/store/store';
import { encodeBase64 } from '@/utils/utils';

import { useAppSelector } from './store-hooks';

interface EndpointParams {
  endpointParam?: string;
  queryParam?: string;
  variablesParam?: string;
  headersParam?: IData[];
}

export const useEncodeUrl = (): {
  replaceUrl: (params?: EndpointParams) => void;
} => {
  const { endpoint, query, variables, headers } = useAppSelector((state: RootState) => state.graphql);
  const locale = useCurrentLocale();

  const replaceHistory = useCallback(
    (urlEnding: string): void => {
      window.history.replaceState(null, '', `/${locale}${ProtectedPaths.GRAPHQL}/${urlEnding}`);
    },
    [locale],
  );

  const getEncodedEndpoint = useCallback(
    (params: EndpointParams = {}): string => {
      const {
        endpointParam = endpoint,
        queryParam = query,
        variablesParam = variables,
        headersParam = headers,
      } = params;

      if (endpointParam) {
        return encodeBase64(endpointParam);
      }

      if (queryParam || variablesParam || headersParam.length !== 0) {
        return NO_ENDPOINT;
      }

      return '';
    },
    [endpoint, query, variables, headers],
  );

  const getEncodedRequestBody = useCallback((): string => {
    const requestBody = JSON.stringify({
      query,
      variables,
    });
    return encodeBase64(requestBody);
  }, [query, variables]);

  const getEncodedHeaders = useCallback(
    (headersParameter: IData[] = headers): string => {
      if (!headersParameter.length) {
        return '';
      }

      const encodedHeaders: string[] = [];
      headersParameter.forEach((header) => {
        encodedHeaders.push(`${header.key}=${encodeURIComponent(header.value)}`);
      });

      return `?${encodedHeaders.join('&')}`;
    },
    [headers],
  );

  const replaceUrl = useCallback(
    (params: EndpointParams = {}): void => {
      const { queryParam = query, variablesParam = variables, headersParam = headers } = params;

      if (queryParam || variablesParam) {
        replaceHistory(`${getEncodedEndpoint(params)}/${getEncodedRequestBody()}${getEncodedHeaders(headersParam)}`);

        return;
      }

      replaceHistory(`${getEncodedEndpoint(params)}${getEncodedHeaders(headersParam)}`);
    },
    [query, variables, headers, getEncodedEndpoint, getEncodedRequestBody, getEncodedHeaders, replaceHistory],
  );

  return {
    replaceUrl,
  };
};
