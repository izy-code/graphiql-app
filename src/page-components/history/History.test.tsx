import { screen } from '@testing-library/react';

import HistoryPage from '@/app/[locale]/history/page';
import { LOCAL_STORAGE_KEY } from '@/hooks/useLocalStorage';
import { renderWithProvidersAndUser } from '@/utils/test-utils';

const translate = vi.fn((arg: string) => {
  switch (arg) {
    case 'empty.title':
      return 'You haven’t executed any requests';
    default:
      return '';
  }
});

vi.mock('next/navigation', async (importOriginal) => {
  const actual = await importOriginal<object>();

  return {
    ...actual,
    useRouter: vi.fn(() => ({ push: vi.fn((arg: string): string => arg) })),
  };
});

vi.mock('@/hooks/useAuth', () => ({ useAuth: vi.fn((): boolean => true) }));

vi.mock('@/locales/client', async (importOriginal) => {
  const actual = await importOriginal<object>();

  return {
    ...actual,
    useScopedI18n: vi.fn(() => translate),
  };
});

describe('History page', () => {
  it('renders correctly', () => {
    const { container } = renderWithProvidersAndUser(<HistoryPage />);

    expect(container).toMatchSnapshot();
  });

  beforeEach(() => {
    localStorage.clear();
  });

  it('show requests list correctly', () => {
    localStorage.setItem(
      LOCAL_STORAGE_KEY,
      JSON.stringify({
        'request-list': ['http://localhost:3000/en/GET/aHR0cHM6Ly9yaWNrYW5kbW9ydHlhcGkuY29tL2FwaS9jaGFyYWN0ZXIvMg/IA'],
      }),
    );
    renderWithProvidersAndUser(<HistoryPage />);

    expect(screen.getByText('https://rickandmortyapi.com/api/character/2')).toBeInTheDocument();
  });

  it('show empty message when no requests', () => {
    renderWithProvidersAndUser(<HistoryPage />);

    expect(screen.getByText('You haven’t executed any requests')).toBeInTheDocument();
  });
});
