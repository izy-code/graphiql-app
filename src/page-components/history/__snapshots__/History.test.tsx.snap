// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`History page > renders correctly 1`] = `
<div>
  <div
    class="_historyPage_3d8cb6"
  >
    <h1
      class="_historyTitle_3d8cb6"
    />
    <div
      class="_noRequests_3d8cb6"
    >
      <h2>
        You havenâ€™t executed any requests
      </h2>
      <p />
      <div
        class="_buttonGroup_3d8cb6"
      >
        <a
          href="/GET"
        />
        <a
          href="/GRAPHQL"
        />
      </div>
    </div>
  </div>
</div>
`;
